### PR TITLE
FIX-#3252: always keep 'by' data in `groupby.__getitem__`

### LINKED ISSUE #3252: `GroupBy.__getitem__` does not include 'by' columns into resulted object which leads to `KeyError`
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Any
- **Modin version** (`modin.__version__`): 
- **Python version**: 3.7.5
- **Code we can use to reproduce**:
```python
import modin.pandas as pd

df = pd.DataFrame(
    {
        "a": [1, 1, 2, 3],
        "b": [3, 4, 5, 6],
        "c": [7, 8, 9, 10],
        "d": [11, 12, 13, 14],
    }
)

grp = df.groupby("a")
grp_view_columns = grp[["b", "c"]]
grp_view_columns.agg(["mean"]) # KeyError: 'a' (failed to default to pandas)

grp = df.groupby(by=["a", [1, 1, 2, 2]])
grp_view_columns = grp[["b", "c"]]
repr(grp_view_columns.mean()) # KeyError: 'a' (modin aggregation is failed)

```

<details><summary>Tracebacks</summary>

First exception:
```
UserWarning: `DataFrame.groupby_on_multiple_columns` defaulting to pandas implementation.
To request implementation, send an email to feature_requests@modin.org.
FutureWarning: The `squeeze` parameter is deprecated and will be removed in a future version.
Traceback (most recent call last):
  File "t3.py", line 14, in <module>
    grp_view_columns.agg(["mean"]) # KeyError: 'a' (failed to default to pandas)
  File "modin/pandas/groupby.py", line 438, in aggregate
    **kwargs,
  File "modin/pandas/groupby.py", line 985, in _default_to_pandas
    return self._df._default_to_pandas(groupby_on_multiple_columns, *args, **kwargs)
  File "modin/pandas/base.py", line 459, in _default_to_pandas
    result = op(pandas_obj, *args, **kwargs)
  File "modin/pandas/groupby.py", line 979, in groupby_on_multiple_columns
    by=by, axis=self._axis, squeeze=self._squeeze, **self._kwargs
  File "miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/frame.py", line 7636, in groupby
    dropna=dropna,
  File "miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 896, in __init__
    dropna=self.dropna,
  File "miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/groupby/grouper.py", line 860, in get_grouper
    raise KeyError(gpr)
KeyError: 'a'
```

Second exception:
```
ray::deploy_ray_func() (pid=1790987, ip=10.125.130.30)
  File "python/ray/_raylet.pyx", line 490, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 497, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 501, in ray._raylet.execute_task
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/ray/util/tracing/tracing_helper.py", line 330, in _function_with_tracing
    return function(*args, **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/ray/pandas_on_ray/frame/axis_partition.py", line 207, in deploy_ray_func
    result = func(*args)
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/base/frame/axis_partition.py", line 303, in deploy_axis_func
    result = func(dataframe, **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/engines/base/frame/data.py", line 1153, in _map_reduce_func
    series_result = func(df, *args, **kwargs)
  File "/localdisk/dchigare/repos/modin_bp/modin/backends/pandas/query_compiler.py", line 2845, in <lambda>
    df, by, drop, partition_idx
  File "/localdisk/dchigare/repos/modin_bp/modin/backends/pandas/query_compiler.py", line 2838, in groupby_agg_builder
    return compute_groupby(df.copy(), drop, partition_idx)
  File "/localdisk/dchigare/repos/modin_bp/modin/backends/pandas/query_compiler.py", line 2742, in compute_groupby
    grouped_df = df.groupby(by=by, axis=axis, **groupby_kwargs)
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/frame.py", line 7636, in groupby
    dropna=dropna,
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 896, in __init__
    dropna=self.dropna,
  File "/localdisk/dchigare/miniconda3/envs/modin_tests/lib/python3.7/site-packages/pandas/core/groupby/grouper.py", line 860, in get_grouper
    raise KeyError(gpr)
KeyError: 'a'
```

</details>

### Describe the problem
The problem that causes these `KeyError`s is that the slice of groupby object (produced by `__getitem__`) contains the source `_df` without "by" columns, only with the cols, specified by the `key` argument:
https://github.com/modin-project/modin/blob/e561a516217abbf630e89f794c838fb878c30a5f/modin/pandas/groupby.py#L353-L360

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

This PR fixes #3252 by always selecting 'by' columns along with the 'key' in `groupby.__getitem__`.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3252 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/groupby.py [hunk 1]
```diff
@@ -954,14 +976,17 @@ def _default_to_pandas(self, f, *args, **kwargs):
...
         ):
             by = self._by.columns[0] if self._drop else self._by.to_pandas().squeeze()
-        elif isinstance(self._by, type(self._query_compiler)):
+        elif self._drop and isinstance(self._by, type(self._query_compiler)):
             by = list(self._by.columns)
         else:
             by = self._by
 
         by = try_cast_to_pandas(by, squeeze=True)
+        by = GroupBy.validate_by(by)
```
Comment (dchigarev) [inline modin/pandas/groupby.py:985]: [`GroupBy.validate_by`](https://github.com/modin-project/modin/blob/4abb53b4a04a66e461a2169f585e474113781be4/modin/data_management/functions/default_methods/groupby_default.py#L34) splits 2D 'by' into a list of 1D Series

File: modin/pandas/groupby.py [hunk 2]
```diff
@@ -954,14 +976,17 @@ def _default_to_pandas(self, f, *args, **kwargs):
             and len(self._by.columns) == 1
         ):
             by = self._by.columns[0] if self._drop else self._by.to_pandas().squeeze()
-        elif isinstance(self._by, type(self._query_compiler)):
+        elif self._drop and isinstance(self._by, type(self._query_compiler)):
```
Comment (dchigarev) [inline modin/pandas/groupby.py:979]: converting QC 'by' to a list of labels column only if this 'by' comes from the `self._df` (if `self._drop is True`)

File: modin/pandas/test/test_groupby.py [hunk 3]
```diff
@@ -439,6 +439,14 @@ def maybe_get_columns(df, by):
     ):
         # Not yet supported for non-original-column-from-dataframe Series in by:
         eval___getattr__(modin_groupby, pandas_groupby, "col3")
+        eval__getitem__(modin_groupby, pandas_groupby, "col3")
+
+    non_by_cols = (
```
Comment (dchigarev) [inline modin/pandas/test/test_groupby.py:444]: cases of intersection of 'by' columns and 'selection' is not supported yet

File: modin/data_management/functions/groupby_function.py [hunk 4]
```diff
@@ -272,6 +273,7 @@ def caller(
             or isinstance(by, pandas.Grouper)
         ):
             by = try_cast_to_pandas(by, squeeze=True)
+            by = GroupBy.validate_by(by)
```
Comment (dchigarev) [inline modin/data_management/functions/groupby_function.py:276]: [`GroupBy.validate_by`](https://github.com/modin-project/modin/blob/4abb53b4a04a66e461a2169f585e474113781be4/modin/data_management/functions/default_methods/groupby_default.py#L34) splits 2D 'by' into a list of 1D Series

File: modin/backends/base/query_compiler.py [hunk 5]
```diff
@@ -2360,7 +2360,7 @@ def groupby_agg(
         """
         if isinstance(by, type(self)) and len(by.columns) == 1:
             by = by.columns[0] if drop else by.to_pandas().squeeze()
-        elif isinstance(by, type(self)):
+        elif drop and isinstance(by, type(self)):
```
Comment (dchigarev) [inline modin/backends/base/query_compiler.py:2363]: converting QC 'by' to a list of column labels only if this 'by' comes from the `self` (if `drop is True`)

File: modin/data_management/functions/groupby_function.py (hunk 4 above)
Comment (gshimansky) [reply modin/data_management/functions/groupby_function.py:276]: It would be really great to have these comments in the code. Nobody reads old closed PRs especially if comments are hidden by "Resolve conversation". All these insights will be lost if they aren't present in the code.

Comment (dchigarev) [reply modin/data_management/functions/groupby_function.py:276]: I just retyped a shorter version of `GroupBy.validate_by` docstring in this comment for reviewers' convenience. In the code, it's your IDE's work to explain what the function is going by docstrings.

Comment (vnlitvinov) [reply modin/data_management/functions/groupby_function.py:276]: I side with @gshimansky here - if you feel that you need to add something so reviewers understand the code better this means that either the code needs to be made cleaner or those comments should be added to the diff. Do not rely on IDE to highlight - there are people who do not use IDEs or do not set up those to show docstrings or whatever.

File: modin/backends/base/query_compiler.py (hunk 5 above)
Comment (vnlitvinov) [reply modin/backends/base/query_compiler.py:2363]: this would make an awesome one-line comment here explaining the "why this is done"

File: modin/pandas/groupby.py [hunk 6]
```diff
@@ -294,6 +296,34 @@ def first(self, **kwargs):
     def backfill(self, limit=None):
         return self.bfill(limit)
 
+    _internal_by_cache = no_default
+
+    # TODO: since python 3.9:
+    # @cached_property
```
Comment (vnlitvinov) [inline modin/pandas/groupby.py:302]: If this is needed in several places let's make a separate PR to introduce our own caching decorator.
I mean it's really easy, like (untested):
```python
if sys.version_info >= (3, 9):
    caching_property = functools.caching_property
else:
    def caching_property(func):
        cache_attr = f"__{func.__name__}_cache"
        @functools.wraps(func)
        def wrap(self, *args, **kw):
            try:
                return getattr(self, cache_attr)
            except AttributeError:
                pass
            value = func(self, *args, **kw)
            setattr(self, cache_attr, value)
            return value
        return wrap
```

File: modin/pandas/groupby.py [hunk 7]
```diff
@@ -294,6 +296,34 @@ def first(self, **kwargs):
...
+    @property
+    def _internal_by(self):
```
Comment (vnlitvinov) [inline modin/pandas/groupby.py:304]: I wonder if a more descriptive name could be made...

File: modin/pandas/groupby.py [hunk 8]
```diff
@@ -294,6 +296,34 @@ def first(self, **kwargs):
...
+        Get only those components of 'by' that are column labels of the source frame.
+
+        Returns
+        -------
+        list of labels
+        """
+        if self._internal_by_cache is not no_default:
+            return self._internal_by_cache
+
+        internal_by = []
```
Comment (vnlitvinov) [inline modin/pandas/groupby.py:315]: can we use a tuple here (as this should be a read-only thing, plus tuples eat less memory)?

File: modin/pandas/groupby.py [hunk 9]
```diff
@@ -954,14 +976,17 @@ def _default_to_pandas(self, f, *args, **kwargs):
...
             by = list(self._by.columns)
         else:
             by = self._by
 
         by = try_cast_to_pandas(by, squeeze=True)
+        by = GroupBy.validate_by(by)
 
         def groupby_on_multiple_columns(df, *args, **kwargs):
+            if isinstance(self, SeriesGroupBy) and isinstance(df, pandas.DataFrame):
+                df = df.squeeze(axis=1)
```
Comment (vnlitvinov) [inline modin/pandas/groupby.py:989]: how is this related to the bug being fixed?

File: modin/pandas/test/test_groupby.py [hunk 10]
```diff
@@ -1111,6 +1119,55 @@ def eval___getattr__(modin_groupby, pandas_groupby, item):
     )
 
 
+def eval__getitem__(md_grp, pd_grp, item):
```
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1122]: Other `eval_*` functions seem to follow the pattern of having three underscores (one from `eval_` and two from a special dunder method), I suggest we keep the pattern.

File: modin/pandas/test/test_groupby.py [hunk 11]
```diff
@@ -439,6 +439,14 @@ def maybe_get_columns(df, by):
     ):
         # Not yet supported for non-original-column-from-dataframe Series in by:
         eval___getattr__(modin_groupby, pandas_groupby, "col3")
+        eval__getitem__(modin_groupby, pandas_groupby, "col3")
+
+    non_by_cols = (
+        [col for col in pandas_df.columns[1:] if col not in modin_groupby._internal_by]
```
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:445]: do we have a separate test for `._internal_by` property?

File: modin/pandas/test/test_groupby.py [hunk 12]
```diff
@@ -1111,6 +1119,55 @@ def eval___getattr__(modin_groupby, pandas_groupby, item):
...
+        pd_grp,
+        lambda grp: grp[item].count(),
+        comparator=build_types_asserter(df_equals),
+    )
+
+    def build_list_agg(fns):
+        def test(grp):
+            res = grp[item].agg(fns)
+            if res.ndim == 2:
+                # Modin's frame has an extra level in the result. Alligning columns to compare.
```
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1140]: This feels like a bug that we work around in the tests. Doesn't feel good.

File: modin/pandas/test/test_groupby.py [hunk 13]
```diff
@@ -1111,6 +1119,55 @@ def eval___getattr__(modin_groupby, pandas_groupby, item):
...
+        build_list_agg(["mean"]),
+        comparator=build_types_asserter(df_equals),
+    )
+    eval_general(
+        md_grp,
+        pd_grp,
+        build_list_agg(["mean", "count"]),
+        comparator=build_types_asserter(df_equals),
+    )
+    # Explicit default-to-pandas test
```
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1159]: What does this test check?

File: modin/pandas/groupby.py (hunk 8 above)
Comment (dchigarev) [reply modin/pandas/groupby.py:315]: We can't do it efficiently, since it's required to do both: appending a single item and extending the container with a sequence, when filling `internal_by` (so no list comprehension here). 

Treating tuple in the same way as a list in a pure for-loops hurts performance, since tuple is immutable, so no "amortized O(1)" insertions here:

<details><summary><i>"Microbenchmarks to the rescue!"</i></summary>

```python
import timeit

N = 1000
M = 10

def list_append(): # timeit() -> 0.41s
    container = list()
    for i in range(N):
        if i % 10:
            container.extend(range(M))
        else:
            container.append(i)

def tuple_append(): # timeit() -> 12.24s
    container = tuple()
    for i in range(N):
        if i % 10:
            container += tuple(range(M))
        else:
            container += (i,)


t1 = timeit.timeit(list_append, number=1000)
t2 = timeit.timeit(tuple_append, number=1000)

print(t1)
print(t2)
```

</details>

We can cast just cast list to a tuple at the end of the function, but will it make sense then?

File: modin/backends/base/query_compiler.py (hunk 5 above)
Comment (dchigarev) [reply modin/backends/base/query_compiler.py:2364]: done

File: modin/data_management/functions/groupby_function.py (hunk 4 above)
Comment (dchigarev) [reply modin/data_management/functions/groupby_function.py:279]: added

File: modin/pandas/test/test_groupby.py (hunk 12 above)
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py:1140]: added a reference to an issue related with this bug

File: modin/pandas/test/test_groupby.py (hunk 11 above)
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py:445]: added a [test](https://github.com/dchigarev/modin/blob/75ae80d64ff790ef2e1d158e31e3c138517408ee/modin/pandas/test/test_groupby.py#L1516) for `._internal_by` at the last commit

File: modin/pandas/test/test_groupby.py (hunk 13 above)
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py:1160]: it checks that the actual bug is gone (#3252), initial report was with default-to-pandas reproducer

File: modin/pandas/groupby.py (hunk 7 above)
Comment (dchigarev) [reply modin/pandas/groupby.py:304]: We already have the "internal_by" term integrated into the groupby flow ([[1]](https://github.com/modin-project/modin/blob/b17db22b04b8d2b9133b64a96a7f62f974d9c959/modin/backends/pandas/query_compiler.py#L2660), [[2]](https://github.com/modin-project/modin/blob/b17db22b04b8d2b9133b64a96a7f62f974d9c959/modin/pandas/dataframe.py#L475-L477), [[3]](https://github.com/modin-project/modin/blob/b17db22b04b8d2b9133b64a96a7f62f974d9c959/modin/backends/pandas/query_compiler.py#L2704-L2707)). Do we need a new term describing the same?

File: modin/pandas/groupby.py (hunk 8 above)
Comment (dchigarev) [reply modin/pandas/groupby.py]: rewritten this function so it uses list-comprehension here (and so using a tuple)

File: modin/pandas/test/test_groupby.py (hunk 10 above)
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py]: aligned

File: modin/pandas/groupby.py (hunk 9 above)
Comment (dchigarev) [reply modin/pandas/groupby.py]: removed as unused code

File: modin/pandas/groupby.py [hunk 14]
```diff
@@ -314,30 +346,20 @@ def __getitem__(self, key):
...
-        # special in which case SeriesGroupBy has to be returned. Such circumstances are when key equals to a single
-        # column name and is not a list of column names or list of one column name.
-        make_dataframe = True
-        if self._drop and self._as_index:
-            if not isinstance(key, list):
-                key = [key]
-                kwargs["squeeze"] = True
+        # The rules of type deduction for the resulted object is the following:
+        #   1. If `key` is a list-like or `as_index is False`, then the resulted object is a DataFrameGroupBy
+        #   2. Otherwise, the resulted object is SeriesGroupBy
```
Comment (dchigarev) [inline modin/pandas/groupby.py:351]: Going from [pandas code](https://github.com/pandas-dev/pandas/blob/5f648bf1706dd75a9ca0d29f26eadfbb595fe52b/pandas/core/base.py#L223-L228)
<details><summary>... and parameters brute-forcing</summary>

```python
df = pandas.DataFrame({"a": [1, 1, 2, 2], "b": [3, 4, 5, 6], "c": [3, 2, 1, 0]})
# self._drop is False for the following 'by':
>>> df.groupby([1, 1, 2, 2])["b"]
<pandas.core.groupby.generic.SeriesGroupBy object at 0x0000017922EF9188>
>>> df.groupby([1, 1, 2, 2], as_index=False)["b"]
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x0000017922EF9108>
>>> df.groupby([1, 1, 2, 2])[["b"]]
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x0000017922EF9088>
# self._drop is True for the following 'by':
>>> df.groupby("a")["b"]
<pandas.core.groupby.generic.SeriesGroupBy object at 0x0000017922EF92C8>
>>> df.groupby("a", as_index=False)["b"]
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x0000017922EF9288>
>>> df.groupby("a")[["b"]]
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x0000017922EF9048>
```

</details>

File: modin/pandas/test/test_groupby.py (hunk 3 above)
Comment (YarShev) [reply modin/pandas/test/test_groupby.py:444]: I suggest adding this comment directly to the code and them removing it after support for cases of the intersection of 'by' columns and 'selection'.

Comment (dchigarev) [reply modin/pandas/test/test_groupby.py:445]: added

File: modin/pandas/groupby.py (hunk 14 above)
Comment (devin-petersohn) [reply modin/pandas/groupby.py:351]: I like these examples, can we add them as comments in the comment here, but replacing pandas with Modin?

Comment (dchigarev) [reply modin/pandas/groupby.py:351]: I'm not sure about that, adding examples makes the comment look more complex than it actually is:
```python
# The rules of type deduction for the resulted object is the following:
#   1. If `key` is a list-like or `as_index is False`, then the resulted object is a DataFrameGroupBy
#   2. Otherwise, the resulted object is SeriesGroupBy
# Examples:
#   - drop: any, as_index: False, __getitem__(key: any) -> DataFrameGroupBy
#   - drop: any, as_index: any, __getitem__(key: list_like) -> DataFrameGroupBy
#   - drop: any, as_index: True, __getitem__(key: label) -> SeriesGroupBy
```
Just a list of rules looks pretty clean and all-covering. What do you think?

Comment (devin-petersohn) [reply modin/pandas/groupby.py:351]: That looks great @dchigarev !

Comment (devin-petersohn) [reply modin/pandas/groupby.py:351]: Would you be able to add the rules there? Thanks!

**Comment by YarShev** [comment]
@dchigarev , LGTM considering @devin-petersohn comments!

**Review by YarShev — APPROVED** [review]

File: modin/pandas/groupby.py (hunk 14 above)
Comment (dchigarev) [reply modin/pandas/groupby.py:358]: added

**Review by devin-petersohn — APPROVED** [review]
Thanks @dchigarev! LGTM
